torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <torch/csrc/distributed/rpc/process_group_agent.h>
+#include <torch/csrc/distributed/rpc/message.h>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+struct FaultyProcessGroupRpcBackendOptions : public ProcessGroupRpcBackendOptions {
+  FaultyProcessGroupRpcBackendOptions(
+      int num_send_recv_threads,
+      std::chrono::milliseconds rpc_timeout,
+      std::string init_method,
+      std::vector<std::string> messages_to_fail,
+      int num_fail_sends = 0)
+      : ProcessGroupRpcBackendOptions(num_send_recv_threads, rpc_timeout, init_method),
+      messagesToFail(messages_to_fail),
+      numFailSends(num_fail_sends) {
+    TORCH_CHECK(numFailSends >= 0, "numFailSends should be non-negative");
+    }
+
+  std::vector<std::string> messagesToFail;
+  int numFailSends;
+};
+
+class FaultyProcessGroupAgent : public ProcessGroupAgent {
```

Comment:
Really like this framework for tests! IIUC, this doesn't need to be just FaultProcessGroupAgent right? It can be FaultyRpcAgent and implement the RpcAgent interface. In the constructor you can take a shared_ptr<RpcAgent>. This way, this agent can be used for any backend and not just ProcessGroupAgent.

If this is too much for this PR, can you create a gh issue to refactor this later to be generic enough for any backend?

Reply:
Sounds good - would be great for testing the robustness of Tensorpipes/Thrfit Agents going forward! Here's the issue (https://github.com/pytorch/pytorch/issues/34754) - happy to work on this after the 1.5 release.